### ISSUE TITLE
channel_append failing without error message when appending more than two channels

### ISSUE THREAD

**Issue by antoche** [OpenImageIO/oiio#2094]
### Problem

I am trying to concatenate some channels from different images using `ImageBufAlgo.channela_append`, and OIIO is failing without telling me much about why. This was originally working on EXRs but failing on JPEGs, but it now fails on both when using 1.9.

OIIO can combine the first two channels, but fails to append the third one, and `geterror()` returns the empty string.

**Expected behavior:** 

`RGB.exr` and `RGB.jpg` are created.

**Actual behavior:**  

On combining  `rgchannels` and `bchannel`, `RuntimeError` is thrown as described in the comments:

    RuntimeError: oiio.ImageBufAlgo.channel_append() failed with error ""

### Steps to Reproduce

    import OpenImageIO as oiio

    def combineChannels( red, green, blue, output ):
        """
        Combine 3 RGB image files into one RGB image file, using the R, G, and B channel
        from the R, G, and B files, respectively.

        Args:
            red (str): path to the file from which to take the R channel
            green (str): path to the file from which to take the G channel
            blue (str): path to the file from which to take the B channel
            output (str): path into which to save the result
        """
        # 1. Read, R,G,B images
        # 2. Retain only R,G,B channel from each buffer, respectively
        # 3. Concatenate channels
        rbuf = oiio.ImageBuf( red )
        rchannel = oiio.ImageBuf()
        assert oiio.ImageBufAlgo.channels(rchannel, rbuf, (0,) )
        
        gbuf = oiio.ImageBuf( green )
        gchannel = oiio.ImageBuf()
        assert oiio.ImageBufAlgo.channels(gchannel, gbuf, (1,) )
        
        # Concatenate R with G
        rgchannels = oiio.ImageBuf()
        assert oiio.ImageBufAlgo.channel_append(rgchannels, rchannel, gchannel )
        
        # Concatenate R+G with B
        rgbchannels = oiio.ImageBuf()
        bbuf = oiio.ImageBuf( blue )
        bchannel = oiio.ImageBuf()
        assert oiio.ImageBufAlgo.channels(bchannel, bbuf, (2,) )
        
        # FIXME: This fails on EXRs with:
        #  - OpenImageIO-1.9.0
        #  - OpenImageIO-1.6.16
        # Works with OpenImageIO-1.5.21
        # FIXME: this fails on JPEGs for all tested versions of OIIO
        if not oiio.ImageBufAlgo.channel_append(rgbchannels, rgchannels, bchannel):
            error = oiio.geterror()
            raise RuntimeError( 'oiio.ImageBufAlgo.channel_append() failed with error "{}"'.format(error) )
        # Write result
        assert rgbchannels.write( output )
        
        
        
    if __name__ == '__main__':
        combineChannels( 'R.exr', 'G.exr', 'B.exr', 'RGB.exr' )
        combineChannels( 'R.jpg', 'G.jpg', 'B.jpg', 'RGB.jpg' )
        
        
### Versions

See above.

Unfortunately, the `OpenImageIO` module does not provide a `__version__` attribute (see https://www.python.org/dev/peps/pep-0396/).

**Comment by lgritz** [comment]
I'll take a look

**Comment by lgritz** [comment]
Given how incredibly old the versions you cite are, I'm not sure if this is 100% helpful, but...

1.9.0 may have been broken, but what was being developed as "1.9" when you took that snapsho actually got released (mere days ago) as 2.0, and it works! (For both exr and jpg.)

So... if upgrading to 2.0 is in the cards for you, whatever was going wrong is long since fixed. And there are lots of other advantages to 2.0, especially if you like using it from Python. The Python bindings have been totally redone with pybind11, spruced up in a number of ways, and are much more "pythonic" -- none of this nonsense about declaring empty ImageBuf's and passing them as output parameters; the IBA functions can all just return ImageBuf results directly. Though, I'm sorry to say, there are a few minor bits that might not be strictly back-compatible, so the jump from 1.8 to 2.0 might take a bit of editing to your source code, though I hope not much.

I'm poking around a bit, bisecting between 1.9.0 and the current head, just to figure out which set of commits fixed this and what was going on. Depending on what it is, it might be something I can backport to 1.8. Stay tuned.

**Comment by antoche** [comment]
Thanks. I haven't seen a build of 2.0 here but I'll check.

**Comment by lgritz** [comment]
Thanks for the clear example, by the way. That really made this hunt a lot easier.

**Comment by lgritz** [comment]
#2095 is a proposal to fix the "always make it float" problem (which wasn't really your problem, just something I noticed along the way).

**Comment by lgritz** [comment]
#2096 is the addition of __version__ to the python module.

**Comment by lgritz** [comment]
It's probably even easier to just force the source images to read as float from the beginning:

    rbuf = oiio.ImageBuf( red )
    rbuf.read (force=True, convert=OpenImageIO.FLOAT)

(Explanation: ImageBuf's from files read automatically when you need them, but you can also force it to happen with a read() call, and that also gives you the option of a data conversion.)

Though I'll be honest, you mentioned a very wide range of OIIO versions, and I'm not sure which ones have the read call take those arguments. Maybe all? Maybe the recent ones? Should be easy to test on your end, though.

**Comment by antoche** [comment]
I did try that, and unfortunately the extra read+convert call didn't seem
to always be heeded (maybe only on JPEGs? I'm not sure anymore). I also
tried copy() with the convert argument, but that wasn't supported in 1.5.

In the end I decided to use the copy() pathway, which works on JPEGs and
EXRs, and stick to 1.9.

**Comment by lgritz** [comment]
Yes, I suspected that the convert argument might have been added more recently (1.5, that's from 2015).

OK, very good. Shall we close this issue?

I just want to remind you again that there is, officially, no 1.9. It was never actually a release and will never have this or anything else backported to create updates. The "1.9" you have is a snapshot of unstable master development at some point, which after many months of further development (and some API change, and total overhaul of the Python binding implementation) eventually became the 2.0 release, which is supported and which will get subsequent minor releases with fixes and enhancements.

If you guys get interested in an upgrade to 2.0, this page explains what you'd need to change on your end: https://github.com/OpenImageIO/oiio/wiki/OIIO-2.0-Porting-Guide
